text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,557 @@
+* Handle user data (Correlations) and observability data (SpanContext, etc) 
+  separately.  
+
+**Extensibility**
+* Allow developers to create new applications for context propagation. For 
+  example: A/B testing, encrypted or authenticated data, and new, experimental 
+  forms of observability.
+
+
+# OpenTelemetry layered architecture
+
+The design of OpenTelemetry is based on the principles of [aspect-oriented 
+programming](https://en.wikipedia.org/wiki/Aspect-oriented_programming), 
+adopted to the needs of distributed systems. 
+
+Some concerns "cut across" multiple abstractions in a program. Logging 
+exemplifies aspect orientation because a logging strategy necessarily affects 
+every logged part of the system. Logging thereby "cross-cuts" across all logged 
+classes and methods. Distributed tracing takes this strategy to the next level, 
+and cross-cuts across all classes and methods in all services in the entire 
+transaction. This requires a distributed form of the same aspect-oriented 
+programming principles in order to be implemented cleanly.
+
+OpenTelemetry approaches this by separating it's design into two layers. The top 
+layer contains a set of independent **cross-cutting concerns**, which intertwine 
+with a program's application logic and cannot be cleanly encapsulated. All 
+concerns share an underlying distributed **context propagation** layer, for 
+storing state and accessing data across the lifespan of a distributed 
+transaction.
+
+
+# Cross-Cutting Concerns
+
+## Observability API
+Distributed tracing is one example of a cross-cutting concern. Tracing code is 
+interleaved with regular code, and ties together independent code modules which 
+would otherwise remain encapsulated. Tracing is also distributed, and requires 
+non-local, transaction-level context propagation in order to execute correctly.
+
+The various observability APIs are not described here directly. However, in this new 
+design, all observability APIs would be modified to make use of the generalized 
+context propagation mechanism described below, rather than the tracing-specific 
+propagation system it uses today.
+
+Note that OpenTelemetry APIs calls should *always* be given access to the entire 
+context object, and never just a subset of the context, such as the value in a 
+single key. This allows the SDK to make improvements and leverage additional 
+data that may be available, without changes to all of the call sites.
+
+## Correlations API
+
+In addition to trace propagation, OpenTelemetry provides a simple mechanism for 
+propagating metric labels, called the Correlations API. Correlations are 
+intended for indexing events in one service with attributes provided by a prior 
+service in the same transaction. This helps to establish a causal relationship 
+between these events. For example, determining that a particular browser version 
+is associated with a failure in an image processing service.
```

Comment:
This is also for inside the same process: one public endpoint (consider you have read/write endpoints) causes failure in a db call (made from the same process).

The point is I want to make sure we say that correlates between processes and inside process between components (e.g. different libraries).

Reply:
I wish you'd remove "metric labels". Correlations are a cross-cutting concern for use in any part of the distributed trace where you wish to have access to this state in real-time, in the process. It's especially useful for metrics, but it's easy to imagine tracing applications.

Reply:
Ok will do. I'm fine saying that Correlations is a completely independent cross-cutting concern, and simply use metrics labels as an example usage.